### DISCUSSION TITLE
wikijs and GIT backup

### DISCUSSION THREAD

**Discussion by Kilthor** [requarks/wiki#5074] [Help / Questions]
Hi I have searched and not found any answers. I have been using Wikijs for about 6 months in Docker using portainer. I have it backing up to github everything was working fine. 3 days ago it stopped backing up and gives this error "error: cannot pull with rebase: Your index contains uncommitted changes. error: please commit or stash them"
1 I have tried using  "git stash " and "git pull -rebase" using CLI in wikijs It just tells me I don't have a repo   /shrug
Is there somewhere with some step by step to work through this ??
I fixed it buy running untracked changes for like the 10th time not sure what is going on but she finally started sycning

**Comment by NGPixel** [comment]
I would suggest using the **Purge Local Repository** action, under Admin > Storage > Git. This will clear the local repository and recreate it. You can then do a **Force Sync**.

**Reply by lecoredump** [reply]
Hello, it's a behaviour I noted multiple times on our instance, there seems to be some cases where Assets as well as new pages end up being staged but not commited, resulting in this message.
@NGPixel, would you have a way / place to see what caused this state (since I have people editing the wiki both in the Git repo as well as from the UI, I can't simply "flush and resync" if I wanna avoid any data loss), to prevent it in the future ?

**Reply by NGPixel** [reply]
You just pointed to the root cause of your issues. You're editing on both GitHub and the wiki at the same time, resulting in merge conflicts.

Either edit on a single platform or ensure you don't edit the same file within the sync window.

The upcoming v3 will have a new direct GitHub integration which will remove the need for a sync window.

**Reply by lecoredump** [reply]
>You just pointed to the root cause of your issues. You're editing on both GitHub and the wiki at the same time, resulting in merge conflicts.

That's not what I described, the "staged" part happens BEFORE the sync even checks for conflicts, on the WIkiJS server itself. The issue is stuff like this : 
![image](https://user-images.githubusercontent.com/30411803/173363407-dc4123ef-1b89-4ca7-b848-da28ed9034da.png)
not being added to any commit and remaining in a "staged" state and messing with any kind of sync (since the "cannot pull with rebase: Your index contains uncommitted changes" doesn't have anything to do with merge conflicts, as it happens before pulling).

I couldn't trace the root cause in the log (couldn't reproduce the behavior as I wanted), but it happens relatively regularly, usually for assets. 

On another note, I'm a little bit afraid about you mentioning "direct Github integration" considering Git storage shouldn't be tied to a specific vendor (or it should be a specific storage for Github itself).

**Reply by NGPixel** [reply]
The git module is not going away. The direct GitHub integration will be a new distinct module.

**Reply by lecoredump** [reply]
Nice to know for the difference between the git and Github modules.
Regarding the Git sync issue, I'll try and narrow it down on an as much as possible  "empty" wiki, since the issue either doesn't seem that common or hasn't been reported yet in the way I described (and create a dedicated issue once I get there).
Anyway, thanks to the whole team, love having a quick (but not dirty) way to use a fully offline version of a wiki.

**Reply by No0key** [reply]
I have same error, using git module and internal gitlab instance as remote storage 
`error: cannot pull with rebase: Your index contains uncommitted changes. error: please commit or stash them. `

**Reply by No0key** [reply]
@lecoredump @NGPixel I found the root cause in my case, it was when i'm trying to load 2 files at the same time. Just try when loading assets select 2 files at once via 'shift' or 'ctrl'.

**Reply by dougbreaux** [reply]
Our instance is in the same state. Can anyone confirm this "purge" process works and is safe? Is there a less extreme process that works?

**Reply by dougbreaux** [reply]
Oh, I missed this
> I fixed it buy running untracked changes

**Reply by dougbreaux** [reply]
No, "Add Untracked Changes" plus "Force Sync" was insufficient to resolve it for me.

**Reply by dougbreaux** [reply]
Nor did "Purge Local" + "Sync" resolve

**Reply by dougbreaux** [reply]
Actually, "Add Untracked" produced this error, but it disappeared too quickly to see it the first time:
> Network Error: JSON.parse: unexpected character at line 1 column 1 of the JSON data

**Reply by No0key** [reply]
@dougbreaux you can fix this by entering couple commands. Login your container in the repo directory and type commands below
- `git add .`
- `git commit -am "fix ..."`

**Reply by dougbreaux** [reply]
@No0key Thanks. I didn't occur to me to shell into the pod and run `git` directly. Which has helped me with some other issues too. The above had indicated there were no uncommitted changes, but deactivating and reactivating the Git Storage instance did give a better error that helped me find _my_ root cause. Which was that the SSH key being used had been removed or expired or something.

**Reply by lecoredump** [reply]
@NGPixel I can confirm No0key's finding about multiple assets uploaded at the same time being linked to this state in the Git repository (although not necessarily through Shift or Ctrl, even selecting multiple files before hitting the "Upload" button can result in this state). 
Still not able to actually pinpoint why it happens, since sometimes everything goes smoothly... Hope this still helps though.

**Reply by CWCarlson123** [reply]
I've had the same issue. Git had not been update for 6 days, local repo was up to date, so just not pushing to git
![image](https://github.com/requarks/wiki/assets/150041949/62778e7f-71a8-44a5-8622-f611bd8ebad4)

I tried @NGPixel fix and didn't work, despite wikijs report the action had been completed and the git storage status was still the same and the Git repo had still not updated (last update 6 days ago). I checked the local repo though the CLI and all was ok. Did a "git status" and said it was ahead by 1 commit, working tree was clean.

I went and had some lunch, came back and (grasping at straws) I clicked the Apply button at the top of the admin > storage page on wikijs... problem solved, git repo last updated an hour ago...huh?! so just before I went on lunch!

so why didn't it clear earlier? why did I have to click apply, after it stated the purge and force sync actions were successful? Them being successful kind of implies that they were applied.

**Reply by serrnovik** [reply]
I'm having the same issue (deployed with current helm version  2.2.20):   

Users edited a few pages. No other pushes from other places:
```python
wikijs-5fc9df9dc5-wxl9p wiki Loading configuration from /wiki/config.yml... OK
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:31:13.045Z [JOB] info: Rendering page ID 1647...
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:31:13.348Z [JOB] info: Using database driver pg for postgres [ OK ]
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:31:14.239Z [JOB] info: Rendering page ID 1647: [ COMPLETED ]
wikijs-5fc9df9dc5-wxl9p wiki Loading configuration from /wiki/config.yml... OK
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:31:14.386Z [JOB] info: Rebuilding page tree...
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:31:14.752Z [JOB] info: Using database driver pg for postgres [ OK ]
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:31:15.118Z [JOB] info: Rebuilding page tree: [ COMPLETED ]
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:31:15.131Z [MASTER] info: (STORAGE/GIT) Committing new file [en] knowledge-base/tools/kimai...
wikijs-5fc9df9dc5-wxl9p wiki Loading configuration from /wiki/config.yml... OK
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:27.191Z [JOB] info: Rebuilding page tree...
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:27.554Z [JOB] info: Using database driver pg for postgres [ OK ]
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:27.921Z [JOB] info: Rebuilding page tree: [ COMPLETED ]
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:27.937Z [MASTER] info: (STORAGE/GIT) Committing file move from [en] user-docs/on-premises/Setting_up_development_environment to [en] dev-docs/Setting_up_development_environment...
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:44.626Z [MASTER] info: Syncing with storage target git...
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:44.631Z [MASTER] info: (STORAGE/GIT) Performing pull rebase from origin on branch main...
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:44.636Z [MASTER] error: Syncing with storage target git: [ FAILED ]
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:32:44.636Z [MASTER] error: error: cannot pull with rebase: You have unstaged changes.
wikijs-5fc9df9dc5-wxl9p wiki error: Please commit or stash them.
wikijs-5fc9df9dc5-wxl9p wiki
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:33:31.288Z [MASTER] info: (STORAGE/GIT) Performing pull rebase from origin on branch main...
wikijs-5fc9df9dc5-wxl9p wiki 2024-04-16T12:33:31.294Z [MASTER] warn: error: cannot pull with rebase: You have unstaged changes.
wikijs-5fc9df9dc5-wxl9p wiki error: Please commit or stash them.
```
![image](https://github.com/requarks/wiki/assets/326432/2bd40d11-d203-46f6-a1ac-a7c921ba8ca4)

I could resolve this by going manually to the pod and:
staging and committing (for some reason *upstaged `.svg` file* that i uploaded *through wikijs ui *(logo) ). 

This is not the first time this happen, but the first time i investigated. Will take a closer look from now.

**Comment by matt1097** [comment]
@NGPixel I also can confirm this bug, and I dug into it a bit. It occurs when uploading multiple assets or sometimes back to back assets updates.

the problem appears to be a race condition. The first asset is still adding/commiting and the git process still has the lock. The second asset then tries to do the same thing, but git fails to grab the lock and fails out.
This leaves the repo with some files added and some unstaged changes which causes the sync process to fail.

I am willing to help out with a PR, but not sure what the best way to approach this just yet.

**Comment by No0key** [comment]
@NGPixel hello. Is there will be bugfix for this issue ?

**Reply by No0key** [reply]
@NGPixel ping

**Comment by No0key** [comment]
@NGPixel

**Comment by matt1097** [comment]
I opened a PR that appears to fix this on my end. https://github.com/requarks/wiki/pull/6511

it is a one line change if anyone wants to test it out in their environment as well.

**Reply by paolofacchinetti** [reply]
hey, just want to say thanks for this! I had to fix the issue through a cron that would sync with the remote git repo, but it was such an ugly hack :')
